### DISCUSSION TITLE
Strange caret behavior

### DISCUSSION THREAD

**Discussion by grimalschi** [codex-team/editor.js#1873] [Q&A]
Hello! When I press "Down" many times, I'm getting strange caret behavior.

What i expect:
<img width="1568" alt="image" src="https://user-images.githubusercontent.com/9027592/61967384-d5233b80-afdd-11e9-9a3f-9d9fc23605d3.png">

What I get:
<img width="1070" alt="image" src="https://user-images.githubusercontent.com/9027592/61967241-8d041900-afdd-11e9-9ab7-b3ebe8fdd89b.png">

Any plans to fix this?

**Comment by hadrysmateusz** [comment]
Any progress on this? It's one of the things that keeps me from using the library as it's super un-intuitive.

**Comment by alexanderisora** [comment]
Any updates on this issue? Thanks.

**Comment by hata6502** [comment]
Absolutely, this issue is valid to improve UI. 
But I have no idea to resolve it.
It may need a DOM hacking.

**Comment by grimalschi** [comment]
In my project I have solved this problem for another block editor.

I am checking every possible caret position and detect closest one for jump.

<img width="688" alt="Screenshot 2020-11-30 at 16 04 18" src="https://user-images.githubusercontent.com/9027592/100613511-b5b0c700-3325-11eb-8bf2-ecc199bd07f3.png">

On picture red is current caret, blue is candidates and green is a closest position for caret.

There is also some optimisation like line break detection, because there may be 2000 lines and its too much to check every possible caret position.

So it is a solution, it works fast enough and I am ready to share some production code and explain this more.

If you (EditorJS team) are interested, write me https://t.me/grimalschi

**Comment by hata6502** [comment]
Great!
By using `Range` and `getBoundingClientRect()`, it may be able to resolve this behavior.

**Comment by hata6502** [comment]
Determining caret position when DOWN key pressed: 

![ezgif com-gif-maker](https://user-images.githubusercontent.com/7702653/100909049-937a9d00-350f-11eb-99c9-73e819e9cd67.gif)

**Comment by neSpecc** [comment]
Awesome! We'll check it

**Comment by hata6502** [comment]
@grimalschi 
@neSpecc 
Sorry! I gave up this PR: https://github.com/codex-team/editor.js/pull/1453

Is there anyone who can resolve this good feature?
I think it is a good feature request.

**Comment by hata6502** [comment]
https://developer.mozilla.org/en-US/docs/Web/API/TreeWalker

**Comment by hata6502** [comment]
Are `force` options in `Caret.navigateNext(force = false)` and `Caret.navigatePrevious(force = false)` used?
I couldn't find where using `force` options. 

```
$ git grep "navigateNext(" src/
src/components/modules/blockEvents.ts:    const isNavigated = navigateNext ? this.Editor.Caret.navigateNext() : this.Editor.Caret.navigatePrevious();
src/components/modules/blockEvents.ts:    const isNavigated = navigatePrevious ? this.Editor.Caret.navigatePrevious() : this.Editor.Caret.navigateNext();
src/components/modules/caret.ts:  public navigateNext(force = false): boolean {

$ git grep "navigatePrevious(" src/
src/components/modules/blockEvents.ts:      if (Caret.navigatePrevious()) {
src/components/modules/blockEvents.ts:    const isNavigated = navigateNext ? this.Editor.Caret.navigateNext() : this.Editor.Caret.navigatePrevious();
src/components/modules/blockEvents.ts:    const isNavigated = navigatePrevious ? this.Editor.Caret.navigatePrevious() : this.Editor.Caret.navigateNext();
src/components/modules/caret.ts:  public navigatePrevious(force = false): boolean {
```

If `force` options are not used, `Caret.navigateNext()` and `Caret.navigatePrevious()` should be simplified by removing `force` options first of all.

**Comment by hata6502** [comment]
I did it! 🎉
Please check this PR:
https://github.com/codex-team/editor.js/pull/1526
https://deploy-preview-1526--editor-js.netlify.app/

**Comment by alexanderjulmer** [comment]
Nice. I found some bugs using Safari (Version 14.0.2 (16610.3.7.1.9)) related to the marker.

**Situation**: Hitting the down arrow key in a line which also contains a marker.
**Expected behavior**: Caret jumps into the next line 
**Actual behavior**: Caret jumps in front of the marker and is stuck there

**Situation**: Hitting the down arrow key while caret is inside a marker.
**Expected behaviour**: Caret jumps into the next line.
**Actual behaviour**:  Caret jumps to the end of the block. Once jumped to the end of the block, the up arrow key doesn't work anymore (left, right and down work).

**Situation**: Hitting the up arrow key while caret is in approx. second half of last line of a block which also contains a marker.
**Expected behaviour**: Caret jumps to second last line.
**Actual behaviour**: Caret jumps to end of block, up arrow key doesn't work anymore.
**Note**: Only happens, if block contains a marker.

Please see attached a video of the first two issues.


https://user-images.githubusercontent.com/10468332/108044741-ef4ebf80-7042-11eb-84ea-112b5bc40d43.mov

**Comment by hata6502** [comment]
@alexanderjulmer 
Thank you for reporting the critical bug on Safari! 
This bug occurs not only https://github.com/codex-team/editor.js/pull/1526 but also `next` branch. 

@neSpecc 
I think this bug should be fixed by the next release.

**Comment by hata6502** [comment]
This bug occurs as `<mark>` element in Safari only.
This caret moving is uncontrolled by Editor.js, and relied on each browser's behavior. 

Umm... I can't know why this bug occurs in `next` branch rather than https://editorjs.io

**Comment by tcper** [comment]
Did this project dead? 
4 years old issue not solved

**Reply by tiagozip** [reply]
I don't think it's dead, It's just not maintained very often
